source/extensions/omni.isaac.lab/test/controllers/test_operational_space.py

```diff
@@ -0,0 +1,864 @@
+            if count % 500 == 0:
+                # check that we converged to the goal
+                if count > 0:
+                    self._check_convergence(osc, ee_pose_b, ee_target_pose_b, ee_force_b, command)
+                # reset joint state to default
+                default_joint_pos = robot.data.default_joint_pos.clone()
+                default_joint_vel = robot.data.default_joint_vel.clone()
+                robot.write_joint_state_to_sim(default_joint_pos, default_joint_vel)
+                robot.set_joint_effort_target(zero_joint_efforts)  # Set zero torques in the initial step
+                robot.write_data_to_sim()
+                robot.reset()
+                # reset contact sensor
+                if self.contact_forces is not None:
+                    self.contact_forces.reset()
+                # reset target pose
+                robot.update(sim_dt)
+                _, _, _, ee_pose_b, _, _, _, _ = self._update_states(
+                    robot, ee_frame_idx, arm_joint_ids
+                )  # at reset, the jacobians are not updated to the latest state
+                command, ee_target_pose_b, ee_target_pose_w, current_goal_idx = self._update_target(
+                    osc, root_pose_w, ee_pose_b, target_set, current_goal_idx
+                )
+                # set the osc command
+                osc.reset()
+                command, task_frame_pose_b = self._convert_to_task_frame(
+                    osc, command=command, ee_target_pose_b=ee_target_pose_b
+                )
+                osc.set_command(
+                    command=command, current_ee_pose_b=ee_pose_b, current_task_frame_pose_b=task_frame_pose_b
+                )
+            else:
+                # get the updated states
+                jacobian_b, mass_matrix, gravity, ee_pose_b, ee_vel_b, root_pose_w, ee_pose_w, ee_force_b = (
+                    self._update_states(robot, ee_frame_idx, arm_joint_ids)
+                )
+                # compute the joint commands
+                joint_efforts = osc.compute(
+                    jacobian_b=jacobian_b,
+                    current_ee_pose_b=ee_pose_b,
+                    current_ee_vel_b=ee_vel_b,
+                    current_ee_force_b=ee_force_b,
+                    mass_matrix=mass_matrix,
+                    gravity=gravity,
+                )
+                robot.set_joint_effort_target(joint_efforts, joint_ids=arm_joint_ids)
+                robot.write_data_to_sim()
+
+            # update marker positions
+            self.ee_marker.visualize(ee_pose_w[:, 0:3], ee_pose_w[:, 3:7])
+            self.goal_marker.visualize(ee_target_pose_w[:, 0:3], ee_target_pose_w[:, 3:7])
+
+            # perform step
+            self.sim.step(render=False)
+            # update buffers
+            robot.update(sim_dt)
+
+    def _update_states(
+        self,
+        robot: Articulation,
+        ee_frame_idx: int,
+        arm_joint_ids: list[int],
+    ):
```

Comment:
please add docstrings for this function

Reply:
"""Update the states of the robot and obtain the relevant quantities for the operational space controller.

        Args:
            robot (Articulation): The robot to control.
            ee_frame_idx (int): The index of the end-effector frame.
            arm_joint_ids (list[int]): The indices of the arm joints.

        Returns:
            jacobian_b (torch.tensor): The Jacobian in the root frame.
            mass_matrix (torch.tensor): The mass matrix.
            gravity (torch.tensor): The gravity vector.
            ee_pose_b (torch.tensor): The end-effector pose in the root frame.
            ee_vel_b (torch.tensor): The end-effector velocity in the root frame.
            root_pose_w (torch.tensor): The root pose in the world frame.
            ee_pose_w (torch.tensor): The end-effector pose in the world frame.
            ee_force_b (torch.tensor): The end-effector force in the root frame.
        """